8.0 Implemented PXB-2608 (Upgrade Vault API to V2)

altmannmarcelo opened the pull request:
https://jira.percona.com/browse/PXB-2608

Adjusted test cases - Splitted test cases into separate file to explore
different secret_mount_point_version options.

Adjusted transition-key size - Vault v2 generates bigger file names (hashed)
which in case of using generate-transition-key would cause vault to
return an error due to file been to big on disk.

Fixed DBUG_ASSRT to assert

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (Initial)

https://jira.percona.com/browse/PS-5364

Initial implementation from Robert.

(cherry picked from commit 74987891fd33f7daedc50c76da6d013ebd66df9d)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (RapidJSON)

https://jira.percona.com/browse/PS-5364

Vault Server responses parsing / generation re-implemented with 'RapidJSON'
library.

Introduced 'Vault_version_type' type instead of plain 'int' that is allowed
to have only a set of predefined values:
  * 'Vault_version_unknown'
  * 'Vault_version_v1'
  * 'Vault_version_v2'.

Fixed 'vault_io-t' and 'vault_parser-t' gunit tests because of the changed
mock interfaces and modified error messages.

(cherry picked from commit 364d3955ab78ae424444dda1e628cf75afbb2dee)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (Major refactoring + gunit)

https://jira.percona.com/browse/PS-5364

Changed the meaning of empty 'secret_mount_point_version' in the plugin
configuration file. It is now considered to be identical to 'AUTO'.

Currently, when this parameter is set to anything rather than '1', plugin
initialization routine needs to perform an additional listing request to the
Hashicorp Vault server ('LIST /v1/sys/mounts/'). In other words:
- if 'secret_mount_point_version' is '1', no 'LIST /v1/sys/mounts/' is
  performed and 'secret_mount_point' is always used as is to form key operation
  URLs
- if 'secret_mount_point_version' is '2', 'LIST /v1/sys/mounts/' is performed
  in order to split 'secret_mount_point' (say 'foo/bar/baz') into
  '<mount_point>' and '<directory>' components (either 'foo/bar' + 'baz' or
  'foo' + 'bar/baz') by iterating over registered mount points - this is
  necessary as in 'kv-v2' secrets URLs have additional 'data' / 'metadata'
  path parts in them ('foo/(meta)?data/bar/baz/...' or
  'foo/bar/(meta)?data/baz/...')
- if 'secret_mount_point_version' is either 'AUTO' or empty (parameter is not
  there at all), 'LIST /v1/sys/mounts/' also needs to be performed in order to
  detect the secret engine type('kv' or 'kv-v2') by looking into 'options'
  subobject of the mount point object corresponding to 'secret_mount_point'
  and checking whether it has 'version' ("1" or "2") in it

Please, notice that listing '/v1/sys/mounts/' requires additional privileges.

Added more validation rules to the plugin configuration parameters:
- 'vault_url' [required] - Vault Server Endpoint URL, must start with either
  'http://' of 'https://'
- 'secret_mount_point' [required] - working path (includes mount point path
   and directory path) of the Vault Server, must not start with '/', must
   not end with '/'
- 'vault_ca' [optional] - path to Vault Server CA certificate, if specified
  'vault_url' must be 'https://'
- 'token' [required] - Vault Server token having enough privileges to
  read/write/delete data in 'secret_mount_point'
- 'secret_mount_point_version' [optional] - Vault Server mount point version,
  must be either '1', '2' or 'AUTO', if this parameter is not specified, it
  is considered to be 'AUTO'

JSON parsing/composing routines extracted into a new 'Vault_parser_composer'
class with 'IVault_parser_composer' interface. The implementation now utilizes
'RapidJSON' library (the same that is used for handling data in standard MySQL
type 'JSON').

Major refactoring in Keyring Vault unit tests:
- configuration files are now generated based on the same 'MTR_VAULT_XXX'
  environment variables as for MTR test cases
- all tests that require real operating Hashicorp Vault server (gunit
  'vault_keyring-api-t', gunit 'vault_io-t', gunit 'vault_io-t' and custom
  'keyring_vault-test') now automatically detect if 'MTR_VAULT_XXX' variables
  are set in the environment and immediately succeed producing a warning if
  they are not
- in gunit tests this is achieved via dynamically modifying '--gtest_filter' -
  all meaningful tests are skipped and only a special test named
'WARNING<test_case_name>_trap.OnlyThisTestIsRunWhenVaultEnvironmentVariablesAreNotSet'
  that unconditionally succeeds is executed
- similarly to 'Mock_logger' class extracted into 'mock_logger.h',
  'Mock_serializer' and 'Mock_serialized_object' extracted into
  'mock_serializer.h' and 'mock_serialized_object.h' respectively as
  they are used in both 'keyring' and 'keyring_vault' gunit tests
- significantly reduced number of library dependencies for the 'keyring_vault'
  unit test executables. Most of them now require only 'gunit_small',
  '${CURL_LIBRARY}' and '${SSL_LIBRARIES}'
- inline/static function definitions from the 'generate_credential_file.h'
  put into proper 'generate_credential_file.cc'
- introduced new 'Vault_environment' class inherited from
  '::testing::Environment' which is responsible for basic per-execution
  'keyring_vault' gunit tests initialization/configuration parameter storage -
  this includes 'curl_global_init()' / 'curl_global_cleanup()', generating
  UUID, generating unique configuration file names and mount point paths
- introduced new 'Vault_test_base' class inherited from '::testing::Test'
  which is used by 'Vault_io_test', 'Vault_keys_container_test' and
  'Keyring_vault_api_test' fixtures as a base in order to automatically
  create/delete configuration file/temporary mount point
- Restored 'keyring_vault-test' custom test which is built under
  'WITH_KEYRING_VAULT_TEST' cmake define. This test simulates heavy load on
  'keyring_vault' plugin by spawning a big number of threads performing random
  (write/read/delete/generate) operations.
- removed 'mysql-test/std_data/keyring_vault_confs/keyring_vault_ut.conf'
  configuration file template as not needed any more

In 'keyring_vault' MTR test suite:
- restored '[mount-empty_v2]' and '[directory-empty_v2]' combinations
- removed remainings of '--source is_vault_server_up.inc'
- re-recorded various test cases because of the new error messages that
  are now being generated from the 'Vault_parser_composer' / 'RapidJSON'.

'IVault_curl' interface extended with 'get_resolved_secret_mount_point_version()'
method. It is now 'Vault_curl''s responsibility to auto-detect / adjust
Vault mount point version.

'IVault_parser_composer' interface in comparison to original 'IVault_parser'
interface also extended with new 'parse_mount_point_version()' method that
is supposed not only auto-detect mount point version but also split
'secret_mount_point' parameter specified in the plugin configuration file
into '<mount_point_path>'(the one actually created as a Vault secret) and
'<directory_path>' (a virtual prefix under '<mount_point_path>'). This is
done based on the result of listing mount points in '/v1/sys/mounts/'.

Completely reworked 'Vault_credentials' class - instead of a key/value
string map we now have strongly-typed fields.

Completely re-implemented 'Vault_credentials_parser' that now reads and
validates parameters from the plugin configuration file and generates
an instance of 'Vault_credentials'.

Fixed parameter const-ness in the 'IKeyring_io::init()' method signature.

Removed various unnecessary header '#include's. Trivial cases reworked with
forward declarations.

(cherry picked from commit 11754b4580d17764649d04de20e59f35b1961345)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (Fwd decl fix)

https://jira.percona.com/browse/PS-5364

Fixed problem with invalid forward declaration of the 'Vault_key' struct.

(cherry picked from commit f8957ef77530bd3ad0ac0e9e2affa9b15da911fa)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (option_labels OOB fix)

https://jira.percona.com/browse/PS-5364

Fixed problem with OOB in 'option_labels' array access.

(cherry picked from commit 6f443acc66ec956ed57f7bb3b65a280014fde92c)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (MP probing)

https://jira.percona.com/browse/PS-5364

Implemented Vault Server mount point secret engine ('kv' / 'kv-v2')
auto-detection mechanism based on '/config' subpath probing.

Now, when 'secret_mount_point' is specified to be 'sub1/sub2/.../subN' and
'secret_mount_point_version' is not explicitly set to '1', we
iterate over
'sub1/config'
'sub1/sub2/config'
...
'sub1/sub2/.../subN/config'
looking for the first subpath in the sequence that returns a valid
HTTP 200 response having 'max_versions', 'cas_required' and
'delete_version_after' in the 'data' subobject.
For instance,
{
  "request_id": "a2c9306a-7f82-6a59-ebfa-bc6142d66c39",
  "lease_id": "",
  "renewable": false,
  "lease_duration": 0,
  "data": {
    "max_versions": 0,
    "cas_required": false,
    "delete_version_after": "0s"
  },
  "wrap_info": null,
  "warnings": null,
  "auth": null
}
If no such subpath found, we consider this mount point to be of version '1'
('kv') and always use 'secret_mount_point_version' as is without splitting to
form key access URLs.
If such subpath 'sub1/sub2/.../subK' found, we consider this mount point to be
of version 2 ('kv-v2') and split 'secret_mount_point' into two components:
- mount point path ('sub1/sub2/.../subK');
- directory path ('subK+1/subK+2/.../subN').
This is necessary because for paths based on 'kv-v2' mount points we need to
know exactly where directory path component starts in order to be able to form
proper key access URLs.
For instance, in the configuration mentioned above, in order to get a key
named 'skey', the following HTTP request should be sent
"GET /v1/sub1/sub2/.../subK/data/subK+1/subK+2/.../subN/skey HTTP/1.1"
Please, notice '/data/' part in the middle of the path here.

This logic is now incorporated into the 'Vault_curl::init()' method.

'IVault_parser_composer' interface / 'Vault_parser_composer' class extended
with new 'parse_mount_point_config()' method which is intended to parse mount
point config responses (like the one mentioned in the example).

Removed 'list_mount_points()' method from the 'IVault_curl' interface /
'Vault_curl' class as it is no longer needed for mount point secret engine
auto-detection.

Changed suppressions in the 'keyring_vault.keyring_vault_config_qa' and
'keyring_vault.keyring_vault_timeout' MTR test cases because of the changed
auto-detection failures error messages.

(cherry picked from commit 3a59da50412250fe730c7c305412332912ed9057)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (MP listing cleanup)

https://jira.percona.com/browse/PS-5364

Removed 'parse_mount_point_version()' method from the 'IVault_parser_composer'
interface / 'Vault_parser_composer' class as it is no longer needed for mount
point secret engine auto-detection.

The following related gunit tests were removed:
- 'Vault_parser_test.GetVaultVersion2';
- 'Vault_parser_test.GetVaultVersion1';
- 'Vault_parser_test.GetVaultOptionsNull';
- 'Vault_parser_test.GetVaultOptionsEmpty'.

Introduced the following new gunit tests to cover mount point config parsing
functionality:
- 'Vault_parser_test.GetMountConfig';
- 'Vault_parser_test.GetMountConfigNull';
- 'Vault_parser_test.GetMountConfigIncomplete'.

(cherry picked from commit d5cdee84342e81b9ddd3ddcc6a6323993ad7b0c9)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (Review comments)

https://jira.percona.com/browse/PS-5364

Addressed review comments.

(cherry picked from commit 92fa6e43af54663ec248f66357f5437991ac1037)

percona-ysorokin committed:
PS-5364 feature: Update Hashicorp Vault plugin to use v2 API (Copyright notice)

https://jira.percona.com/browse/PS-5364

Added Copyright notice to all 'keyring_vault' source file, including gunit tests.

(cherry picked from commit c66e186dee46a02e547ea1b5fe9244a4c32195fe)

percona-ysorokin committed:
PS-7589 fix: keyring_vault Secrets Engine Version auto-detection mechanism fails on Vault Server < 1.5.0

https://jira.percona.com/browse/PS-7589

Fixed problem with Secrets Engine Version auto-detection mechanism when
'/config' probing was expecting a JSON subobject containing
'delete_version_after' key. This parameter was added only in the Vault
Server 1.5.0.

Fixed by making this 'delete_version_after' key optional during response
parsing.

'Vault_parser_test.GetMountConfig' gunit test extended with an additional
payload.

(cherry picked from commit 7e3059eedb9b90c1fc25ab8615d6a931ce98139e)

Commit:
PS-5364 fix: Update Hashicorp Vault plugin to use v2 API (fix keyring_vault-test test)

https://jira.percona.com/browse/PS-5364
(cherry picked from commit 327d24c6b614372e6ad44810f922616c907dd869)

altmannmarcelo committed:
Implemented PXB-2608 (Upgrade Vault API to V2)

https://jira.percona.com/browse/PXB-2608

Adjusted test cases - Splitted test cases into separate file to explore
different secret_mount_point_version options.

Adjusted transition-key size - Vault v2 generates bigger file names (hashed)
which in case of using generate-transition-key would cause vault to
return an error due to file been to big on disk.

Fixed DBUG_ASSRT to assert

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtrabackup?pullRequest=1179) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: altmannmarcelo<br/>:x: oleksandr-kachan<br/>:x: percona-ysorokin<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtrabackup?pullRequest=1179) it.</sub>

altmannmarcelo commented:
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/116/ - keyring file component/both are an issue in the way boot the server, not related to this PR.

Local Vault tests:
```
 VAULT_URL=https://192.168.23.44:8200 VAULT_CACERT=${HOME}/vault.crt VAULT_CA=${HOME}/vault.crt VAULT_TOKEN=s.aKCe6mbqQ6Zx9FHcaDYlqgR9 ./run.sh -s keyring
Using /work/pxb/ins/8.0/xtrabackup-test/test/var as test root
Detecting server version...
Running against Percona Server 8.0.25-15-debug (XtraDB 8.0.25-15)
Using 'xtrabackup' as xtrabackup binary
Autodetected number of cores: 32
Limiting to 16 to avoid excessive resource consumption
Using 16 parallel workers
Running suite(s): keyring

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
bug1590351                               w1	[passed]    34
innodb_keyring_proxy_plugin              w9	[passed]    38
innodb_encryption_mix_plugins            w4	[passed]    50
keyring_encryption                       w16	[passed]    78
innodb_encryption_export                 w3	[passed]    90
pxb-1705                                 w4	[passed]    116
pxb-1793                                 w16	[passed]    88
innodb_redo_log_encrypt                  w15	[passed]    173
keyring_pxb_2275                         w1	[passed]    204
pxb-1903                                 w4	[passed]    71
page_compression                         w9	[passed]    200
pxb-1879                                 w3	[passed]    157
pxb-1937                                 w15	[passed]    79
pxb-1936                                 w16	[passed]    131
pxb-1954                                 w1	[passed]    80
pxb-2429                                 w4	[passed]    91
innodb_keyring_vault                     w10	[passed]    353
innodb_keyring_vault_auto_v2             w12	[passed]    353
innodb_keyring_vault_v1                  w13	[passed]    353
innodb_keyring_vault_v2                  w14	[passed]    353
shared_tablespace_encryption             w3	[passed]    120
innodb_keyring_plugin_to_component       w8	[passed]    367
innodb_keyring_vault_auto_v1             w11	[passed]    367
reencrypt                                w9	[passed]    143
system_tablespace_encryption             w15	[passed]    133
bug1737525                               w2	[passed]    401
xb_log_archiving_encrypt                 w16	[passed]    132
innodb_keyring_file_plugin               w7	[passed]    467
innodb_keyring_file_both                 w5	[passed]    488
innodb_keyring_file_component            w6	[passed]    519
==============================================================================
Spent 6229 of 533 seconds executing testcases

SUMMARY: 30 run, 30 successful, 0 skipped, 0 failed

See results/ for detailed output
```

satya-bodapati commented on storage/innobase/xtrabackup/src/keyring_plugins.cc:
```diff
@@ -200,23 +205,23 @@ static bool xb_fetch_key(char *key_name, char *key) {
 static bool xb_create_transition_key(char *key_name, char *key) {
-  byte rand32[32];
-  char rand64[45];
+  byte rand32[20];
```
shouldn't this be 32 bytes? it seems the intention is create 64byte random key/prefix?

satya-bodapati replied:
How will this work vault v1? How will PXB read keys which are in v1 vault version?

altmannmarcelo replied:
This is not a problem.
This is a Key=>value secret we store on vault. It could be anything. we then store it on xtrabackup_keys:
```
strings /tmp/backup/xtrabackup_keys
KEYSV01W=
XBKey-c70f89d3-1af1-11ec-9b69-d45d64347a19-tUTpWs/F+vwY1VJTgZdeeOKFEsA=
}~y
SrYb
}~y
SrYb
```
Here `XBKey-c70f89d3-1af1-11ec-9b69-d45d64347a19-tUTpWs/F+vwY1VJTgZdeeOKFEsA=` is the the key, and the last part `tUTpWs/F+vwY1VJTgZdeeOKFEsA=` is what we generated. At prepare we will read this key from xtrabackup_keys file and fetch it from keyring (vault or file). This is just stamp we add on each backup to avoid colisions. We are still generating an 30 ascii char len sufix key, which will be more than enough for backups&increments to not colide.

For example, the server implements it by using an incremental integer as suffix.

satya-bodapati replied:
ok, the readers don't insist the key to be of specific len and pattern. Got it.

satya-bodapati commented on storage/innobase/xtrabackup/test/inc/keyring_vault.sh:
```diff
@@ -39,16 +39,27 @@ function keyring_vault_ping()
 
 function keyring_vault_mount()
 {
+	local VAULT_CONFIG_VERSION=$1
```
CONFIG_VERSION is unused here

altmannmarcelo replied:
good catch. Fixed.

satya-bodapati commented on storage/innobase/xtrabackup/src/keyring_plugins.cc:
```diff
@@ -158,8 +158,13 @@ dberr_t xb_set_encryption(fil_space_t *space) {
+const size_t TRANSITION_KEY_PREFIX_LEN =
+    sizeof(TRANSITION_KEY_PREFIX) / sizeof(char);
+const size_t TRANSITION_KEY_RANDOM_DATA_LEN = 32;
+const size_t TRANSITION_KEY_NAME_MAX_LEN = TRANSITION_KEY_PREFIX_LEN +
+                                           Encryption::SERVER_UUID_LEN +
+                                           TRANSITION_KEY_RANDOM_DATA_LEN + 1;
```
Dont we need the length of v1 for reading and writing keys to v1? Should we instead create TRANSITION_KEY_NAME_MAX_LEN_V1 and _V2?

altmannmarcelo replied:
This is not necessary, as I mentioned on https://github.com/percona/percona-xtrabackup/pull/1179#discussion_r715532010 this is just a stamp to make the key unique, all previous backups (with longer key) have the key saved at xtrabackup_keys file and preparing with a new version will be able to read that key and fetch vault to retrieve the secret necessary to un-encrypt tables. This is `TRANSITION_KEY_NAME_MAX_LEN` is only used on generating new transition key code path. Not used at prepare.

satya-bodapati replied:
Thanks for clarifying marce.

xb_tablespace_keys_load_one() creates a buffer of TRANSITION_KEY_NAME_MAX_LEN. This will be 13 chars less now. How do we deal with loading keys into the buffer of smaller size ?

altmannmarcelo replied:
@satya-bodapati you are correct. We do need versioning on the transition key, otherwise, using PXB 8.0.27 we won't be able to read transition keys prior to this version (longer keys). I've adjusted that.

altmannmarcelo committed:
Added key v2 and fixed test

Added transition key v2 to be able to differ from v1 (longer key) and
v2 (smaller key).

Added missing version option on test cases.

satya-bodapati commented on storage/innobase/xtrabackup/src/keyring_plugins.cc:
```diff
@@ -508,17 +519,19 @@ bool xb_tablespace_keys_load_one(const char *dir, const char *transition_key,
-
-  if (fread(transition_key_name, 1, sizeof(transition_key_name), f) !=
-      sizeof(transition_key_name)) {
+  transition_key_name =
+      (char *)my_malloc(PSI_NOT_INSTRUMENTED, transition_key_name_size,
+                        MYF(MY_FAE | MY_ZEROFILL));
```
possible to used std::unique_ptr?

satya-bodapati replied:
or the variant with custom deleter (we can use my_free() there)

altmannmarcelo replied:
@satya-bodapati good suggestion. I've added smart pointer. 

Regarding the 2.4, I will do the changes there too. However, since the 5.7 and 8.0 commits related to vault v2 are different, I've not merged the 2.4 branch into the 8.0. I had to cherry pick them individually. I will do a null merge after both branches are merged.

satya-bodapati reviewed (changes requested):
Looks good marce! just one minor suggestion. Also, I forgot to start reviewing from base PR (2.4) Can you please incorporate comments on the 2.4 branch as well?

altmannmarcelo committed:
Added smart pointer

Adjusted transition_key_name pointer to use std::unique_ptr.

satya-bodapati reviewed (approved):
LGTM!
